File: GFG/Greedy/jobseq.py
class Job:
    def __init__(self, id, deadline, profit):
        self.id = id
        self.deadline = deadline
        self.profit = profit

def job_sequencing(jobs):
    n = len(jobs)
    result = ['-'] * n  # stores the final job sequence
    slot = [False] * n  # keeps track of available time slots

    jobs.sort(key=lambda x: x.profit, reverse=True)  # sort jobs by profit in descending order

    for i in range(n):
        for j in range(min(n, jobs[i].deadline) - 1, -1, -1):
            if not slot[j]:
                result[j] = jobs[i].id
                slot[j] = True
                break

    # Print the job sequence
    print(' '.join(job for job in result if job != '-'))

jobs = [Job('a', 2, 100), Job('b', 1,50), Job('c', 2,10),
        Job('d', 1,20), Job('e', 3,30)]

job_sequencing(jobs)



# arr=[(4,70),(1,80),(1,30),(1,100)]
# arr=[(2,50),(2,60),(3,20),(3,30)]
# arr=[(2,100),(1,50),(2,10),(1,20),(3,30)]
# print(arr) 

File: GFG/Greedy/jobseq.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Job {
    char id;
    int deadline;
    int profit;

    Job(char i, int d, int p) : id(i), deadline(d), profit(p) {}
};

bool sortByProfit(Job j1, Job j2) {
    return (j1.profit > j2.profit);
}

void jobSequencing(vector<Job>& jobs) {
    int n = jobs.size();
    vector<char> result(n, '-'); // stores the final job sequence
    vector<bool> slot(n, false); // keeps track of available time slots

    sort(jobs.begin(), jobs.end(), sortByProfit); // sort jobs by profit in descending order

    for (int i = 0; i < n; i++) {
        for (int j = min(n, jobs[i].deadline) - 1; j >= 0; j--) {
            if (!slot[j]) {
                result[j] = jobs[i].id;
                slot[j] = true;
                break;
            }
        }
    }

    // Print the job sequence
    for (int i = 0; i < n; i++) {
        if (result[i] != '-')
            cout << result[i] << " ";
    }
    cout << endl;
}

int main() {
    vector<Job> jobs = { {'a', 2, 100}, {'b', 1, 19}, {'c', 2, 27},{'d', 1, 25}, {'e', 3, 15} };

    jobSequencing(jobs);

    return 0;
}
